### PR TITLE
[opentitantool] Fixes to bootstrap of legacy Titan

### PR DESCRIPTION
**Pull request by jesultra**
This newly added code has a few bugs.

1. A bug was introduced during review, when the code to set the
"final" flag on the last frame was move out of the loop.  The hash of
the block needs to be computed after the flag is set.  (This bug would
have been caught had I run the one and only unit test in this file.)

2. The above lead to the discovery that the existing logic (also in
primitive.rs) does not attempt to verify the ack of the very last
block, leading to the false claim of successful bootstrapping.

3. Further real-life testing revealed more shortcomings of the
existing logic.  Ordinarily, the bidirectional transaction is used
such that the boot rom will ack the previously transmitted frame, at
the same time as receiving the next one.  I.e. the Rust code will look
for ack of frame 3 while transmitting frame 4.  If an unrecognized
checksum is received, the code will go back and retransmit frame 3
repeatedly, until ack of frame 3 is received.  However, if the cause
of the original bad ack was not in the boot rom, but due to noise on
the CIDO line, then the bootrom may have successfully received frame
3, and possibly also frame 4.  If the latter is the case, the boot rom
(on H1Dx at least) will repeat the ack of block 4 while frame 3 is
being retransmitted.  This CL introduces code to recognize this case,
and others, rather than only looking for ack of the frame most
recently transmitted.

TEST=Repeatedly flashed H1Dx while manipulating CIDO capacitance

Signed-off-by: Jes B. Klinke <jbk@chromium.org>

### CODE REVIEW

File: sw/host/opentitanlib/src/bootstrap/legacy.rs [hunk 1]
```diff
@@ -202,40 +206,74 @@ impl UpdateProtocol for Legacy {
...
-                    break;
-                }
-                log::error!(
-                    "Frame hash mismatch device:{:x?} != frame:{:x?}.",
-                    &response[..Frame::HASH_LEN],
-                    want_hash
-                );
-                // Retry sending same frame.
+            if response[..Frame::HASH_LEN] == frames[first_unacked_index].frame_hash() {
+                first_unacked_index = first_unacked_index + 1;
```
Comment (mcy) [inline sw/host/opentitanlib/src/bootstrap/legacy.rs:260]: This can be `+= 1;`.

**Review by mcy — APPROVED** [review]

File: sw/host/opentitanlib/src/bootstrap/legacy.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/legacy.rs:260]: I considered that, but on the other hand, I like the symmetry of having both if statements follow the same pattern:
```
if ... frames[X]... {
  ... = X + 1;
} ...
```

File: sw/host/opentitanlib/src/bootstrap/legacy.rs [hunk 2]
```diff
@@ -202,40 +206,74 @@ impl UpdateProtocol for Legacy {
...
+
+        // Counts the number of repeated failures in getting a frame ack.
+        let mut consecutive_errors = 0;
+
+        // Set if the past transaction ack'ed the block sent before that one, while transmitting
+        // the next.  If so, we heuristically assume that the block just sent will be ack'ed in
+        // the upcoming bidirectional transaction.
+        let mut optimistic = false;
+        
+        // Set if the past transaction ack'ed the block seet before that one, while
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/bootstrap/legacy.rs:221]: s/seet/seen/ or s/seet/sent/ ?

**Review by cfrantz — APPROVED** [review]

File: sw/host/opentitanlib/src/bootstrap/legacy.rs [hunk 3]
```diff
@@ -202,40 +206,74 @@ impl UpdateProtocol for Legacy {
...
+        // re-transmitting the same block.  If so, we do not care about the ack in the upcoming
+        // bidirectional transaction.
+        let mut becoming_optimistic = false;
+
+        loop {
+            if consecutive_errors > 100 {
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/legacy.rs]: Done

File: sw/host/opentitanlib/src/bootstrap/legacy.rs (hunk 2 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/bootstrap/legacy.rs]: I have elaborated a little on this comment.

**Review by cfrantz — APPROVED** [review]

**Pull request merged**